raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java

```diff
@@ -0,0 +1,489 @@
+
+    @Test
+    void testVotersRecordEndpointsWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        Iterator<VotersRecord.Endpoint> iterator = endpoints.votersRecordEndpoints();
+
+        assertFalse(iterator.hasNext());
+    }
+
+    @Test
+    void testSizeWithOneEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        int size = endpoints.size();
+
+        assertEquals(1, size);
+    }
+
+    @Test
+    void testSizeWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        int size = endpoints.size();
+
+        assertEquals(0, size);
+    }
+
+    @Test
+    void testIsEmptyWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertFalse(endpoints.isEmpty());
+    }
+
+    @Test
+    void testIsEmptyWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        assertEquals(0, endpoints.size());
+    }
+
+    @Test
+    void testEqualsAndHashCodeWithSameEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        Endpoints sameEndpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertEquals(endpoints, sameEndpoints);
+        assertEquals(endpoints.hashCode(), sameEndpoints.hashCode());
+    }
+
+    @Test
+    void testEqualsAndHashCodeWithDifferentEndpoints() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        Map<ListenerName, InetSocketAddress> anotherEndpointMap = new HashMap<>();
+        anotherEndpointMap.put(
+                ListenerName.normalised("another"),
+                InetSocketAddress.createUnresolved("localhost", 9093)
+        );
+        Endpoints differentEndpoints = Endpoints.fromInetSocketAddresses(anotherEndpointMap);
+
+        assertNotEquals(endpoints, differentEndpoints);
+        assertNotEquals(endpoints.hashCode(), differentEndpoints.hashCode());
+    }
+
+    @Test
+    void testToString() {
```

Comment:
What the value of testing `toString`? what public interfaces will `toString` impact if it change to any other string?

Reply:
Was thinking it is a good idea to make sure it conforms to the overridden format. Happy to remove it if needed.

Reply:
I don't see `toString` has any impact to be tested in the context of this class. We do test `toString` for example for metrics related code because it has an impact but this one not that much. If you checked the code you will see that we don't actively testing `toString`

Reply:
removed